### ISSUE TITLE
Knowledge to use UBOs and SSBOs for materials

### ISSUE THREAD

**Issue by psilord** [HackerTheory/first-light#15]
Materials should use UBOs and SSBOs instead of straight up uniforms. Here is some
wisdom from mfiano about this:

```
TL;DR instructions

<mfiano> 1) bind a shader block of a specific program to an int, using
	 BIND-SHADER-STORAGE-BLOCK. 2) create a buffer, passing it this same
	 int, using CREATE-BUFFER. 3) write data whenever with
	 WRITE-BUFFER-PATH.

Long instructions:

<mfiano> 1) First, you're going to want to view shadow's internal state
	 global, shadow::*shader-info* and get the name of the
	 program. assuming it's a keyword symbol named :program, I'll
	 continue...
<mfiano> 2) Then you're going to want to get the block name too. Let's call it
	 :block for simplicity. Both of these could very well be symbols in
	 another non-keyword package, but I'm being simple here  [21:51]
<mfiano> 3) So we have a program name and a block name in that program we want
	 to bind to a buffer. next we bind that block to an integer, called a
	 binding point: (bind-shader-storage-block :program :block 10). here
	 10 is arbitrary. It can be anything from 0 to max binding points
	 allowed by your hardware.  [21:53]
<mfiano> These 3 steps take care of binding a shader block to a binding point,
	 but we also have to do the other side....bind a gpu buffer to the
	 same binding point
<mfiano> This linking allows the shader to read any data written to from the
	 cpu...either once, or as often as you like  [21:54]
<mfiano> Anyway for the other side:
<mfiano> (create-buffer :ssbo :name 10). Here, :name can be anything, and 10
	 must be the same 10 as before. What this does is 2 things: It first
	 creates a fresh buffer allocating storage according to the buffer
	 that is already bound to the binding point 10. This means you MUST
	 bind a block before you bind a buffer. A buffer needs to know the
	 block layout in order to figure out its size etc  [21:56]
<mfiano> So now you have a buffer called :NAME that has the same size and
	 layout your shader expects. All that's left is to write to that
	 buffer for you shader to see any changes made to it  [21:57]
<psilord> And do we have a nice way of writing into that buffer?  [21:58]
<mfiano> That is done by: (write-buffer-path :name some-field some-data). Here
	 :name is the name of the previously created buffer. some-field is a
	 keyword denoting the actual defstruct-gpu field name in the shader,
	 and data is any scalar, vec, mat, or array thereof, corresponding to
	 the type it expects.  [21:59]
<mfiano> Hierarchical fields are separated with a dot: (write-buffer-path
	 :name :a.b.c (v3:make 1.0 2.0 3.0)) Would write a vec3 to the c slot
	 of the b struct of the a struct, as an example.  [22:00]
<psilord> So yeah, mostly we do have a way.
<mfiano> It's actually a very simple 3 steps, but it has to be done nicely in
	 that binding points start at zero and are only reused when needed
	 etc.  [22:02]
<mfiano> This is so similar to active texture units its not even funny
```

**Comment by psilord** [comment]
There is a piece of test code in the core materials file that implements the fundamental idea. But, it needs thought for how to make it better, plus maybe some additional features for shadow.

**Comment by psilord** [comment]
mfiano and I have been making progress on this. Slowly but surely it'll get completed.
